README.md
Language: [English](README.md) | [中文简体](README-ZH.md)
# tween_image_widget

A tween image widget Flutter package.

## Getting Started

### Add dependency

```yaml
dependencies:
  tween_image_widget: ^0.0.2 #please use the latest version on pub
```
 
### Use Widget
1. Use repeat animation
 Only need to pass in the image’s path and index, and the specific durationMilliseconds.
```dart
         TweenImageWidget(
              ImagesEntry(1, 8, "assets/img/addpage_icon_load%s.png"),
              durationMilliseconds: 500,
            ),
```

2. Use control animation
Need to assign repeat to false, define TweenImageWidget, and then use reverse(), forward(), stop() and other functions according to the scene
```dart
  @override
  void initState() {
    _curtainAnimationImage = TweenImageWidget(
      ImagesEntry(1, 20, "equipmentcontrol_pic_curtain%s".toAssetImg()),
      durationMilliseconds: 5000,
      repeat: false,
    );
    super.initState();
  }
```

3. Other used parameters (initial value, width, height)
```dart
///initial value:calculated by percentage
startsValue: curtainPosition / 100,
///Height and width are optional. Once assigned, the image will be scaled according to the value you specify
height: 50,
width: 50,
```

**Functions:**
```dart
  ///[repeat]'s default value is true, if need manually control the animation, please pass in false
  ///
  ///[startsValue] Is a starting state value, if want it to start from a few percent of the animation, please pass in this value
  ///
  ///[durationMilliseconds] is how many milliseconds the animation to run, the default value is 3000 milliseconds
  TweenImageWidget(
    this.entry, {
    this.width,
    this.height,
    this.durationMilliseconds = 3000,
    this.repeat = true,
    this.startsValue,
    key,
  }

  ///[lowIndex] is the first index of the animation sequence pictures
  /// [highIndex] is the last index of the animation sequence pictures
  /// [basePath] is the file path of the animation sequence pictures,example:"images/pic_animation%s.png"
  /// [makePath] is an optional file path constructor
  ImagesEntry(this.lowIndex, this.highIndex, this.basePath, [this.makePath]);

  
 ///Stops running this animation.
  stop({bool canceled = true});

  ///Starts running this animation forwards (towards the end).
  forward({double? from});

  ///Starts running this animation in reverse (towards the beginning).
  reverse({double? from});

  ///Drives the animation from its current value to target.
  animateToTarget(double target, {Duration? duration});
```

### Example:

```dart
 bool isDoorOpen = true;
  TweenImageWidget _doorAnimationImage;

  @override
  void initState() {
    _doorAnimationImage = TweenImageWidget(ImagesEntry(1, 4, "images/equipmentcontrol_pic_door%s"),
        durationMilliseconds: 600, repeat: false, startsValue: isDoorOpen ? 1 : 0);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Expanded(child: _doorAnimationImage),
        Column(
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                InkWell(
                  child: Text("Lock"),
                  onTap: () {
                    _doorAnimationImage.stop();
                    _doorAnimationImage.reverse();
                  },
                ),
                InkWell(
                  child: Text("Unlock"),
                  onTap: () {
                    _doorAnimationImage.stop();
                    _doorAnimationImage.forward();
                  },
                ),
              ],
            ),
          ],
        )
      ],
    );
  }
```

#### Demo Pictrue
![demo](demo.gif)






CHANGELOG.md
## [0.0.2] - Release date: 2021-03-29.

* Second version: Improve documentation and small optimizations.
## [0.0.1] - Release date: 2021-03-15.

* First version: initial release.


test/tween_image_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:tween_image_widget/tween_image_widget.dart';

void main() {
  test('TODO ', () {});
}


lib/tween_image_widget.dart
library tween_image_widget;

import 'package:flutter/material.dart';

/// Tween image widget 💖💖💖
/// Widget that can play the animation repeatedly, or manually control the progress of the animation sequence
///
///example:
///```dart
///TweenImageWidget(
///       ImagesEntry(1, 25, "images/pic_animation%s.png"),
///       durationMilliseconds: 5000,
///       repeat: false,
///       startsValue: 0.2,//animationValue / 100
///     )
/// ```
///
// ignore: must_be_immutable
class TweenImageWidget extends StatefulWidget {
  /// Starts running this animation in the forward direction, and
  /// restarts the animation when it completes.
  ///
  ///Reference on [AnimationController.repeat]
  final bool repeat;
  final double? width;
  final double? height;
  final ImagesEntry entry;
  final int durationMilliseconds;
  AnimationController? _animationController;

  ///Reference on [AnimationController.value],is should be >= 0.0 && <= 1.0
  ///
  ///Because of [animationController] default value is lowerBound = 0.0, upperBound = 1.0,
  final double? startsValue;

  ///Get AnimationController
  AnimationController? get animationController {
    return _animationController;
  }

  ///Stops running this animation.
  ///
  ///Reference on [AnimationController.stop]
  void stop({bool canceled = true}) =>
      _animationController?.stop(canceled: canceled);

  ///Starts running this animation forwards (towards the end).
  ///
  ///Reference on [AnimationController.forward]
  TickerFuture? forward({double? from}) =>
      _animationController?.forward(from: from);

  ///Starts running this animation in reverse (towards the beginning).
  ///
  ///Reference on [AnimationController.reverse]
  TickerFuture? reverse({double? from}) =>
      _animationController?.reverse(from: from);

  ///Drives the animation from its current value to target.
  ///
  ///Reference on [AnimationController.animateTo] and [AnimationController.animateBack]
  TickerFuture? animateToTarget(double target, {Duration? duration}) {
    if (target > (_animationController?.value ?? 0)) {
      return _animationController?.animateTo(target, duration: duration);
    } else {
      return _animationController?.animateBack(target, duration: duration);
    }
  }

  ///[repeat]'s default value is true, if need manually control the animation, please pass in false
  ///
  ///[startsValue] Is a starting state value, if want it to start from a few percent of the animation, please pass in this value
  ///
  ///[durationMilliseconds] is how many milliseconds the animation to run, the default value is 3000 milliseconds
  TweenImageWidget(
    this.entry, {
    this.width,
    this.height,
    this.durationMilliseconds = 3000,
    this.repeat = true,
    this.startsValue,
    key,
  })  : assert(
            (width == null || width > 0), "width is should be null or > 0.0"),
        assert((height == null || height > 0),
            "height is should be null or > 0.0"),
        assert((startsValue == null || (startsValue >= 0 && startsValue <= 1)),
            "startsValue is should be null or >= 0.0 && <= 1.0"),
        super(key: key);

  @override
  _TweenImageWidgetState createState() =>
      _TweenImageWidgetState((animationController) {
        _animationController = animationController;
      });
}

class _TweenImageWidgetState extends State<TweenImageWidget>
    with TickerProviderStateMixin {
  AnimationController? _controller;
  late Animation<int> _animation;
  void Function(AnimationController? animationController)
      setAnimationController;

  _TweenImageWidgetState(this.setAnimationController);

  @override
  void initState() {
    super.initState();

    _controller = new AnimationController(
        vsync: this,
        duration: Duration(milliseconds: widget.durationMilliseconds));

    if (widget.startsValue != null) {
      _controller!.value = widget.startsValue!;
    }

    if (widget.repeat) {
      _controller!.repeat();
    }

    _animation =
        new IntTween(begin: widget.entry.lowIndex, end: widget.entry.highIndex)
            .animate(_controller!);

    setAnimationController(_controller);
  }

  @override
  Widget build(BuildContext context) {
    return new AnimatedBuilder(
      animation: _animation,
      builder: (BuildContext context, Widget? child) {
        int frame = _animation.value;
        return new Image.asset(
          (widget.entry.makePath ?? widget.entry._defaultMakePath()).call(
              frame), //根据传进来的参数拼接路径 Splice the path according to the parameters passed in
          gaplessPlayback: true, //避免图片闪烁 avoid flickering pictures
          width: widget.width,
          height: widget.height,
          fit: (widget.width != null || widget.height != null)
              ? BoxFit.fill
              : null,
        );
      },
    );
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }
}

class ImagesEntry {
  ///First index of pictures
  int lowIndex = 0;

  ///Last index of pictures
  int highIndex = 0;

  ///File path of pictures,%s will be replaced by the index
  String basePath;

  ///File path constructor
  String Function(int index)? makePath;

  ///Default file path constructor
  String Function(int index) _defaultMakePath() =>
      (index) => basePath.replaceAll("%s", index.toString());

  ///[lowIndex] is the first index of the animation sequence pictures
  ///
  /// [highIndex] is the last index of the animation sequence pictures
  ///
  /// [basePath] is the file path of the animation sequence pictures,example:"images/pic_animation%s.png"
  ///
  /// [makePath] is an optional file path constructor
  ImagesEntry(this.lowIndex, this.highIndex, this.basePath, [this.makePath]);
}


example/lib/device_curtain_page.dart
import 'package:flutter/material.dart';
import 'package:tween_image_widget/tween_image_widget.dart';

class DeviceCurtainPage extends StatefulWidget {
  DeviceCurtainPage({Key key}) : super(key: key);

  @override
  _DeviceCurtainPageState createState() => _DeviceCurtainPageState();
}

class _DeviceCurtainPageState extends State<DeviceCurtainPage> {
  ///0 ~ 100
  int curtainPosition = 50;

  bool isOnline = true;
  TweenImageWidget _curtainAnimationImage;

  @override
  void initState() {
    _curtainAnimationImage = TweenImageWidget(
      ImagesEntry(1, 20, "equipmentcontrol_pic_curtain%s".toAssetImg()),
      durationMilliseconds: 5000,
      repeat: false,
      startsValue: curtainPosition / 100,
    );
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        //equipmentcontrol_curtain_off
        Expanded(child: _curtainAnimationImage),
        Container(
          padding: EdgeInsets.symmetric(vertical: 42),
          color: MyTheme.color_bg_white,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: [
              buildTurnWidget(
                  "TurnOn",
                  isOnline
                      ? "equipmentcontrol_open_on"
                      : "equipmentcontrol_open_offline", () {
                setState(() {
                  _curtainAnimationImage.stop();
                  _curtainAnimationImage.reverse();
                });
              }),
              buildTurnWidget(
                  "Pause",
                  isOnline
                      ? "equipmentcontrol_pause_on"
                      : "equipmentcontrol_stop_disabled", () {
                setState(() {
                  _curtainAnimationImage.stop();
                });
              }),
              buildTurnWidget(
                  "TurnOff",
                  isOnline
                      ? "equipmentcontrol_close_on"
                      : "equipmentcontrol_close_offline", () {
                setState(() {
                  _curtainAnimationImage.stop();
                  _curtainAnimationImage.forward();
                });
              }),
            ],
          ),
        ),
      ],
    );
  }

  Column buildTurnWidget(
      final String name, final String imageUrl, GestureTapCallback onTap) {
    return Column(
      children: [
        InkWell(
          child: Card(
            shape: CircleBorder(side: BorderSide.none),
            elevation: 2,
            shadowColor: MyTheme.color_bg_black,
            child: CircleAvatar(
              radius: 24,
              backgroundColor:
                  isOnline ? MyTheme.color_bg_orange : MyTheme.color_bg_main,
              child: Image.asset(imageUrl.toAssetImg()),
            ),
          ),
          onTap: onTap,
        ),
        SizedBox(height: 6),
        Text(
          name,
          style: MyTheme.style_subContent,
        ),
      ],
    );
  }
}

class MyTheme {
  MyTheme._();

  static const Color color_bg_white = Color(0xFFFFFFFF);
  static const Color color_bg_black = Color(0xFF000000);
  static const Color color_bg_orange = Color(0xFFF6B402);
  static const Color color_bg_main = Color(0xFFF9F6EF);

  static const Color color_text_main = Color(0xFF000000);

  static const style_subContent =
      TextStyle(color: MyTheme.color_text_main, fontSize: 12);
}

extension StringEx on String {
  String toAssetImg() {
    return "assets/img/${this}.png";
  }
}


example/lib/main.dart
import 'package:example/device_curtain_page.dart';
import 'package:flutter/material.dart';
import 'package:tween_image_widget/tween_image_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'tween_image_widget Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Tween Image Widget Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(height: 20),
            Text(
              'Demo 1: repeat animation',
              style: Theme.of(context).textTheme.headline6,
            ),
            SizedBox(height: 20),
            TweenImageWidget(
              ImagesEntry(1, 8, "assets/img/addpage_icon_load%s.png"),
              durationMilliseconds: 500,
              height: 50,
              width: 50,
            ),
            SizedBox(height: 30),
            Text(
              'Demo 2: control animation',
              style: Theme.of(context).textTheme.headline6,
            ),
            Expanded(
              child: DeviceCurtainPage(),
            ),
          ],
        ),
      ),
    );
  }
}
